examples/notebooks/python/criteo-etl-ratio.ipynb

```diff
@@ -0,0 +1,504 @@
+    "            # because if we try to do it in one query spark ends up assigning the\n",
+    "            # partial ids in two different locations that are not guaranteed to line up\n",
+    "            # this prevents that from happening by assigning the partial ids\n",
+    "            # and then writeing them out.\n",
+    "            save_low_mem_partial_ids(\n",
+    "                    assign_low_mem_partial_ids(col_counts),\n",
+    "                    model_folder,\n",
+    "                    write_mode)\n",
+    "            save_combined_model(\n",
+    "                    assign_low_mem_final_ids(load_low_mem_partial_ids(spark, model_folder)),\n",
+    "                    model_folder,\n",
+    "                    write_mode)\n",
+    "            save_column_models(\n",
+    "                get_column_models(load_combined_model(spark, model_folder)),\n",
+    "                model_folder,\n",
+    "                write_mode)\n",
+    "\n",
+    "        \n",
+    "    if mode == 'transform':\n",
+    "        spark.conf.set('spark.sql.shuffle.partitions', day_range*apply_shuffle_parallel_per_day)\n",
+    "        with _timed('transform'):\n",
+    "            print(\"transform\")\n",
+    "            origin_df = origin_df.withColumn('ordinal', monotonically_increasing_id())\n",
+    "\n",
+    "            models = list(load_column_models(spark, model_folder, False))\n",
+    "            models = [(i, df, agg.sum, flag) for i, df, agg, flag in models]\n",
+    "          \n",
+    "            df = apply_models(origin_df,models,False,1.0)\n",
+    "            df = transform_log(df, False)\n",
+    "            df = df.orderBy('ordinal')\n",
+    "            df = df.drop('ordinal')\n",
+    "            df = df.drop('day')\n",
+    "\n",
+    "            print(\"writing final output\")\n",
+    "            df.write.parquet(\n",
+    "                output_folder,\n",
+    "                mode=write_mode)\n",
+    "\n",
+    "    print('=' * 100)\n",
+    "    print(_benchmark)\n",
+    "\n",
+    "\n",
+    "if __name__ == '__main__':\n",
+    "    conf = SparkConf().setAppName('ETL')\\\n",
+    "            .set('spark.executor.memory', '20g')\\\n",
+    "            .set('spark.executor.resource.gpu.amount', 1)\\\n",
+    "            .set('spark.executor.cores', 2)\\\n",
+    "            .set('spark.cores.max', 2)\\\n",
+    "            .set('spark.task.cpus', 1)\\\n",
+    "            .set('spark.task.resource.gpu.amount', 0.5)\\\n",
+    "            .set('spark.rapids.sql.concurrentGpuTasks', 1)\n",
+    "\n",
+    "    spark = SparkSession.builder.config(conf=conf).getOrCreate()\n",
+    "    # generate model for cat columns\n",
+    "    _main(spark, \"generate models\")\n",
+    "    # transform data with model and save to new parquet\n",
+    "    _main(spark, \"transform\")\n",
+    "    spark.stop()"
+   ]
+  },
+  {
+   "cell_type": "code",
```

Comment:
nit: there're 3 empty cells at the end of this notebook

Reply:
Done